File: src/contracts.zig
//! Closed, versioned capture contracts and typed outcomes.
//!
//! The wire payload is one JSON object. The schema is closed: unknown fields,
//! duplicate fields, missing fields, and over-budget values are all rejected
//! with typed reasons rather than best-effort acceptance.

const std = @import("std");

pub const payload_schema_version: u32 = 1;
pub const episode_schema = "aj-episode.v1";

pub const max_payload_bytes: usize = 4 * 1024 * 1024;
pub const max_content_bytes: usize = 2 * 1024 * 1024;
/// Read budget for one rendered episode file: the file carries frontmatter
/// and escaping overhead on top of payload content, so it may legitimately
/// exceed `max_payload_bytes`.
pub const max_episode_file_bytes: usize = max_payload_bytes * 2;
/// Directory descent budget below the journal root. The deepest supported
/// layout is worlds/<world>/scopes/<scope>/lanes/<lane>/YYYY/MM/DD/file.
pub const corpus_walk_depth: u8 = 10;
pub const max_world_len: usize = 64;
pub const max_token_len: usize = 128;
pub const max_tools: usize = 256;

pub const Lane = enum {
    conversation,
    delegated_work,
    evaluation,
    imported_legacy,
};

/// Capture result vocabulary reported to adapters. `published` and
/// `duplicate` are success; everything else is a distinct typed failure.
pub const CaptureOutcome = enum {
    published,
    duplicate,
    conflict,
    malformed,
    permission_denied,
    unavailable,
    internal_error,
};

/// Index freshness is reported independently of source publication.
pub const IndexFreshness = enum {
    fresh,
    stale,
    not_built,
    unavailable,
};

/// Retrieval outcome vocabulary shared by `memory_search` and `memory_get`.
/// `no_match` is a valid result, not a failure: empty, stale, unavailable,
/// malformed, and timed-out are all distinct typed outcomes.
pub const Outcome = enum {
    match,
    no_match,
    stale_revision,
    gone,
    index_stale,
    timeout,
    unavailable,
    permission_denied,
    malformed,
    conflict,
    internal_error,
};

// Retrieval bounds. Search returns ranked references and bounded snippets;
// `memory_get` opens one bounded span. Neither ever returns an unbounded
// episode body.
pub const max_query_bytes: usize = 4096;
pub const max_query_terms: usize = 64;
pub const max_snippet_line_bytes: usize = 400;
pub const max_snippet_bytes: usize = 4096;
pub const max_get_lines: u32 = 400;
pub const max_get_bytes: usize = 64 * 1024;
pub const max_results_limit: u32 = 100;

/// Wire shape prior to validation. Field names are the wire contract.
/// `world` and `scope` may be omitted and filled from owner defaults. A
/// first-class host may provide them when transporting an explicit owner
/// session selection. Every other field is a lifecycle fact the adapter
/// knows.
pub const RawPayload = struct {
    schema_version: u32,
    world: ?[]const u8 = null,
    scope: ?[]const u8 = null,
    lane: []const u8,
    harness: []const u8,
    adapter_version: []const u8,
    session_id: []const u8,
    turn_id: []const u8,
    event_time_ms: u64,
    capture_policy: []const u8,
    turn_outcome: []const u8,
    user_content: []const u8,
    assistant_result: []const u8,
    tools: ?[]const Tool = null,

    pub const Tool = struct {
        name: []const u8,
    };
};

/// Validated payload. Slices alias the arena inside `std.json.Parsed`; keep
/// the parse result alive for as long as the payload is used.
pub const Payload = struct {
    world: []const u8,
    scope: []const u8,
    lane: Lane,
    harness: []const u8,
    adapter_version: []const u8,
    session_id: []const u8,
    turn_id: []const u8,
    event_time_ms: u64,
    capture_policy: []const u8,
    turn_outcome: []const u8,
    user_content: []const u8,
    assistant_result: []const u8,
    tools: []const RawPayload.Tool,
};

pub const ValidateError = error{
    UnsupportedSchemaVersion,
    InvalidWorld,
    InvalidScope,
    InvalidLane,
    InvalidHarness,
    InvalidAdapterVersion,
    InvalidSessionId,
    InvalidTurnId,
    InvalidCapturePolicy,
    InvalidTurnOutcome,
    EmptyUserContent,
    EmptyAssistantResult,
    OversizedContent,
    InvalidUtf8,
    TooManyTools,
    InvalidToolName,
};

pub const ParseError = ValidateError || error{ Malformed, OutOfMemory };

/// A world id names a directory component: lowercase alphanumeric plus `-`,
/// bounded, never starting with `.` (enforced by charset).
pub fn validWorld(s: []const u8) bool {
    if (s.len == 0 or s.len > max_world_len) return false;
    for (s) |c| switch (c) {
        'a'...'z', '0'...'9', '-' => {},
        else => return false,
    };
    return true;
}

/// Identity tokens (session, turn, harness, policy, scope, outcome, adapter
/// version): printable, no whitespace/control bytes, so they embed safely in
/// frontmatter lines and canonical digest input.
pub fn validToken(s: []const u8) bool {
    if (s.len == 0 or s.len > max_token_len) return false;
    for (s) |c| switch (c) {
        'a'...'z', 'A'...'Z', '0'...'9', '.', '_', '-', ':', '+', '/', '@' => {},
        else => return false,
    };
    return true;
}

/// Scope names are both frontmatter tokens and directory components. Unlike
/// general identity tokens they cannot contain a path separator or name a
/// traversal component.
pub fn validScope(s: []const u8) bool {
    if (!validToken(s) or std.mem.eql(u8, s, ".") or std.mem.eql(u8, s, ".."))
        return false;
    return std.mem.indexOfScalar(u8, s, '/') == null;
}

pub fn validate(raw: RawPayload) ValidateError!Payload {
    if (raw.schema_version != payload_schema_version) return error.UnsupportedSchemaVersion;
    const world = raw.world orelse return error.InvalidWorld;
    const scope = raw.scope orelse return error.InvalidScope;
    if (!validWorld(world)) return error.InvalidWorld;
    if (!validScope(scope)) return error.InvalidScope;
    const lane = std.meta.stringToEnum(Lane, raw.lane) orelse return error.InvalidLane;
    if (!validToken(raw.harness)) return error.InvalidHarness;
    if (!validToken(raw.adapter_version)) return error.InvalidAdapterVersion;
    if (!validToken(raw.session_id)) return error.InvalidSessionId;
    if (!validToken(raw.turn_id)) return error.InvalidTurnId;
    if (!validToken(raw.capture_policy)) return error.InvalidCapturePolicy;
    if (!validToken(raw.turn_outcome)) return error.InvalidTurnOutcome;
    if (raw.user_content.len == 0) return error.EmptyUserContent;
    if (raw.assistant_result.len == 0) return error.EmptyAssistantResult;
    if (raw.user_content.len > max_content_bytes) return error.OversizedContent;
    if (raw.assistant_result.len > max_content_bytes) return error.OversizedContent;
    if (!std.unicode.utf8ValidateSlice(raw.user_content)) return error.InvalidUtf8;
    if (!std.unicode.utf8ValidateSlice(raw.assistant_result)) return error.InvalidUtf8;
    const tools = raw.tools orelse &[_]RawPayload.Tool{};
    if (tools.len > max_tools) return error.TooManyTools;
    for (tools) |t| if (!validToken(t.name)) return error.InvalidToolName;
    return .{
        .world = world,
        .scope = scope,
        .lane = lane,
        .harness = raw.harness,
        .adapter_version = raw.adapter_version,
        .session_id = raw.session_id,
        .turn_id = raw.turn_id,
        .event_time_ms = raw.event_time_ms,
        .capture_policy = raw.capture_policy,
        .turn_outcome = raw.turn_outcome,
        .user_content = raw.user_content,
        .assistant_result = raw.assistant_result,
        .tools = tools,
    };
}

/// Parses and validates one capture payload. The returned `Parsed` owns every
/// slice in the `Payload`; deinit it when publication is complete.
pub fn parsePayload(
    gpa: std.mem.Allocator,
    bytes: []const u8,
) ParseError!std.json.Parsed(RawPayload) {
    if (bytes.len > max_payload_bytes) return error.Malformed;
    return std.json.parseFromSlice(RawPayload, gpa, bytes, .{
        .allocate = .alloc_always,
    }) catch |err| switch (err) {
        error.OutOfMemory => error.OutOfMemory,
        else => error.Malformed,
    };
}

pub const test_payload_json =
    \\{
    \\  "schema_version": 1,
    \\  "world": "testworld",
    \\  "scope": "workspace:demo",
    \\  "lane": "conversation",
    \\  "harness": "claude-code",
    \\  "adapter_version": "0.1.0",
    \\  "session_id": "sess-01",
    \\  "turn_id": "turn-0007",
    \\  "event_time_ms": 1783862400123,
    \\  "capture_policy": "default-v1",
    \\  "turn_outcome": "completed",
    \\  "user_content": "How do the naïve tests behave? — ✓",
    \\  "assistant_result": "They pass.\n\n```zig\nconst x = 1;\n```",
    \\  "tools": [{"name": "Bash"}, {"name": "Read"}]
    \\}
;

test "valid payload parses and validates" {
    const parsed = try parsePayload(std.testing.allocator, test_payload_json);
    defer parsed.deinit();
    const p = try validate(parsed.value);
    try std.testing.expectEqual(Lane.conversation, p.lane);
    try std.testing.expectEqualStrings("testworld", p.world);
    try std.testing.expectEqual(@as(usize, 2), p.tools.len);
}

test "unknown field is rejected as malformed" {
    const bad =
        \\{"schema_version":1,"world":"w","scope":"s","lane":"conversation",
        \\"harness":"h","adapter_version":"1","session_id":"a","turn_id":"b",
        \\"event_time_ms":1,"capture_policy":"p","turn_outcome":"ok",
        \\"user_content":"u","assistant_result":"a","surprise":true}
    ;
    try std.testing.expectError(error.Malformed, parsePayload(std.testing.allocator, bad));
}

test "missing required field is rejected as malformed" {
    const bad =
        \\{"schema_version":1,"world":"w"}
    ;
    try std.testing.expectError(error.Malformed, parsePayload(std.testing.allocator, bad));
}

test "invalid lane and world are typed validation failures" {
    const parsed = try parsePayload(std.testing.allocator, test_payload_json);
    defer parsed.deinit();
    var raw = parsed.value;
    raw.lane = "gossip";
    try std.testing.expectError(error.InvalidLane, validate(raw));
    raw = parsed.value;
    raw.world = "Bad World";
    try std.testing.expectError(error.InvalidWorld, validate(raw));
    raw = parsed.value;
    raw.assistant_result = "";
    try std.testing.expectError(error.EmptyAssistantResult, validate(raw));
    raw = parsed.value;
    raw.user_content = "\xff\xfe not utf8";
    try std.testing.expectError(error.InvalidUtf8, validate(raw));
}

test "oversized content is a typed validation failure" {
    const gpa = std.testing.allocator;
    const parsed = try parsePayload(gpa, test_payload_json);
    defer parsed.deinit();
    var raw = parsed.value;
    const big = try gpa.alloc(u8, max_content_bytes + 1);
    defer gpa.free(big);
    @memset(big, 'x');
    raw.user_content = big;
    try std.testing.expectError(error.OversizedContent, validate(raw));
}

test "omitted world and scope parse as null and fail validation unfilled" {
    const bare =
        \\{"schema_version":1,"lane":"conversation","harness":"pi",
        \\"adapter_version":"1","session_id":"s","turn_id":"t",
        \\"event_time_ms":1,"capture_policy":"p","turn_outcome":"completed",
        \\"user_content":"u","assistant_result":"a"}
    ;
    const parsed = try parsePayload(std.testing.allocator, bare);
    defer parsed.deinit();
    try std.testing.expect(parsed.value.world == null);
    try std.testing.expect(parsed.value.scope == null);
    try std.testing.expectError(error.InvalidWorld, validate(parsed.value));

    // The capture host's config merge makes the same payload valid.
    var filled = parsed.value;
    filled.world = "main";
    filled.scope = "default";
    const p = try validate(filled);
    try std.testing.expectEqualStrings("main", p.world);
    try std.testing.expectEqualStrings("default", p.scope);
}

File: adapters/pi/engine/contracts.ts
// Closed, versioned capture contracts and typed outcomes.
//
// The wire payload is one JSON object. The schema is closed: unknown fields,
// duplicate fields, missing fields, and over-budget values are all rejected
// with typed reasons rather than best-effort acceptance. Ported from the Go
// engine's contracts.go; the acceptance is a frozen Interface-tier contract.

import { parseOrderedJson, objectGet, type JsonValue, type JsonEntry } from "./json.ts";

/** The only capture wire version accepted. */
export const PAYLOAD_SCHEMA_VERSION = 1;

/** Stamps every rendered episode's frontmatter. */
export const EPISODE_SCHEMA = "aj-episode.v1";

// Size and shape budgets.
export const MAX_PAYLOAD_BYTES = 4 * 1024 * 1024;
export const MAX_CONTENT_BYTES = 2 * 1024 * 1024;
// The read budget for one rendered episode file: the file carries
// frontmatter and separator overhead on top of payload content, so it may
// legitimately exceed MAX_PAYLOAD_BYTES.
export const MAX_EPISODE_FILE_BYTES = MAX_PAYLOAD_BYTES * 2;
// Bounds directory descent below the journal root. The deepest supported
// layout is worlds/<world>/scopes/<scope>/lanes/<lane>/YYYY/MM/DD/file.
export const CORPUS_WALK_DEPTH = 10;
export const MAX_WORLD_LEN = 64;
export const MAX_TOKEN_LEN = 128;
export const MAX_TOOLS = 256;
// Bounds the optional provenance paths (workspace root, branch-of).
export const MAX_PATH_LEN = 512;

// Retrieval bounds. Search returns ranked references and bounded snippets;
// get opens one bounded span. Neither ever returns an unbounded episode body.
export const MAX_QUERY_BYTES = 4096;
export const MAX_QUERY_TERMS = 64;
export const MAX_SNIPPET_LINE_BYTES = 400;
export const MAX_SNIPPET_BYTES = 4096;
export const MAX_GET_LINES = 400;
export const MAX_GET_BYTES = 64 * 1024;
export const MAX_RESULTS_LIMIT = 100;

/**
 * Lane distinguishes normal conversation, delegated work, evaluation, and
 * explicit imported legacy source. A system record type, never a user
 * folder choice.
 */
export type Lane = "conversation" | "delegated_work" | "evaluation" | "imported_legacy";

export const LANES: readonly Lane[] = ["conversation", "delegated_work", "evaluation", "imported_legacy"];

export function validLane(l: string): l is Lane {
  return (LANES as readonly string[]).includes(l);
}

// Event times outside this window are refused at validate with
// ImplausibleEventTime: a wrapped or garbage timestamp would otherwise
// shard the episode into a nonsense date directory. The bounds are
// deliberately wide — the epoch through 9999-12-31T23:59:59Z — because the
// contract's job is refusing nonsense, not judging clocks.
export const MIN_EVENT_TIME_MS = 0n;
export const MAX_EVENT_TIME_MS = 253402300799000n;

/**
 * The capture result vocabulary reported to adapters. published and
 * duplicate are success; everything else is a distinct typed failure.
 * Consumers must tolerate values they do not know: the vocabulary is an
 * interface-tier contract. (v2 removed `superseded`: a redelivery either
 * matches the recorded digest — duplicate — or conflicts.)
 */
export type CaptureOutcome =
  | "published"
  | "duplicate"
  | "conflict"
  | "malformed"
  | "permission_denied"
  | "unavailable"
  | "internal_error";

/** Index freshness, reported independently of source publication. */
export type IndexFreshness = "fresh" | "stale" | "not_built" | "unavailable";

/**
 * The retrieval vocabulary shared by memory_search and memory_get.
 * no_match is a valid result, not a failure.
 */
export type Outcome =
  | "match"
  | "no_match"
  | "stale_revision"
  | "gone"
  | "index_stale"
  | "timeout"
  | "unavailable"
  | "permission_denied"
  | "malformed"
  | "conflict"
  | "internal_error";

/**
 * Validation failure vocabulary, carried in a capture report's `detail`.
 * These CamelCase codes are an Interface-tier contract adapters match on.
 * parsePayload throws only Malformed; validate throws one specific code so
 * the CLI maps each to its contract outcome without string matching.
 */
export type CaptureErrorCode =
  | "Malformed"
  | "UnsupportedSchemaVersion"
  | "InvalidWorld"
  | "InvalidScope"
  | "InvalidLane"
  | "InvalidHarness"
  | "InvalidAdapterVersion"
  | "InvalidSessionId"
  | "InvalidTurnId"
  | "InvalidCapturePolicy"
  | "InvalidTurnOutcome"
  | "ImplausibleEventTime"
  | "EmptyUserContent"
  | "EmptyAssistantResult"
  | "OversizedContent"
  | "InvalidUtf8"
  | "TooManyTools"
  | "InvalidToolName"
  | "InvalidWorkspaceRoot"
  | "InvalidBranchOf"
  | "InvalidHost"
  | "ContainmentViolation"
  | "PermissionDenied"
  | "Unavailable";

export class CaptureError extends Error {
  readonly code: CaptureErrorCode;
  constructor(code: CaptureErrorCode) {
    super(code);
    this.name = "CaptureError";
    this.code = code;
  }
}

/** Maps any thrown value to the capture error vocabulary. */
export function captureErrorName(err: unknown): CaptureErrorCode {
  return err instanceof CaptureError ? err.code : "Unavailable";
}

/** The allowlisted safe metadata for one tool call: its name only. */
export interface Tool {
  readonly name: string;
}

/**
 * The wire shape prior to validation. world and scope may be omitted and
 * filled from owner defaults. eventTimeMs stays a bigint here so an
 * out-of-range uint64 is carried exactly to the implausibility check.
 */
export interface RawPayload {
  schemaVersion: number;
  world: string | null;
  scope: string | null;
  lane: string;
  harness: string;
  adapterVersion: string;
  sessionId: string;
  turnId: string;
  eventTimeMs: bigint;
  capturePolicy: string;
  turnOutcome: string;
  userContent: string;
  assistantResult: string;
  tools: Tool[] | null; // null when the wire omitted the field
  // Optional session provenance: where the turn happened. Excluded from the
  // payload digest like other capture-source metadata, so a faithful
  // re-delivery still dedupes.
  workspaceRoot: string | null;
  branchOf: string | null;
  host: string | null;
}

/** A validated capture payload. */
export interface Payload {
  world: string;
  scope: string;
  lane: Lane;
  harness: string;
  adapterVersion: string;
  sessionId: string;
  turnId: string;
  eventTimeMs: number;
  capturePolicy: string;
  turnOutcome: string;
  userContent: string;
  assistantResult: string;
  tools: Tool[];
  workspaceRoot: string | null;
  branchOf: string | null;
  host: string | null;
}

const utf8ByteLength = (s: string): number => Buffer.byteLength(s, "utf8");

/**
 * Whether s names a directory component: lowercase alphanumeric plus '-',
 * bounded, never starting with '.' (enforced by charset).
 */
export function validWorld(s: string): boolean {
  if (s.length === 0 || s.length > MAX_WORLD_LEN) return false;
  return /^[a-z0-9-]+$/.test(s);
}

/**
 * Whether s is a safe identity token (session, turn, harness, policy,
 * scope, outcome, adapter version): printable, no whitespace or control
 * bytes, so it embeds safely in frontmatter lines and canonical digest
 * input. The charset is ASCII, so code-unit length equals byte length.
 */
export function validToken(s: string): boolean {
  if (s.length === 0 || utf8ByteLength(s) > MAX_TOKEN_LEN) return false;
  return /^[A-Za-z0-9._\-:+/@]+$/.test(s);
}

/**
 * Whether s is safe as a frontmatter line value. Provenance paths are never
 * directory components or digest input, so the rule is line safety, not a
 * charset: bounded, well-formed, and free of control bytes. Spaces and
 * non-ASCII are legitimate in real filesystem paths and are allowed.
 */
export function validPath(s: string): boolean {
  if (s.length === 0 || utf8ByteLength(s) > MAX_PATH_LEN) return false;
  if (!s.isWellFormed()) return false;
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    if (c < 0x20 || c === 0x7f) return false;
  }
  return true;
}

/**
 * Whether s is a usable scope name. Scopes are both frontmatter tokens and
 * directory components; unlike general identity tokens they cannot contain
 * a path separator, name a traversal component, or start with '.' — the
 * corpus walk skips dot-directories as foreign tooling state, so a dot-led
 * scope would publish episodes the walk could never see.
 */
export function validScope(s: string): boolean {
  if (!validToken(s) || s.startsWith(".")) return false;
  return !s.includes("/");
}

// The closed wire object.
const REQUIRED_KEYS = [
  "schema_version",
  "lane",
  "harness",
  "adapter_version",
  "session_id",
  "turn_id",
  "event_time_ms",
  "capture_policy",
  "turn_outcome",
  "user_content",
  "assistant_result",
] as const;

const OPTIONAL_KEYS = new Set(["world", "scope", "tools", "workspace_root", "branch_of", "host"]);

const malformed = () => new CaptureError("Malformed");

/**
 * Parses the wire bytes into a RawPayload. Every parse-level problem — over
 * budget, invalid JSON, duplicate or unknown fields, missing required
 * fields, wrong value types — collapses to Malformed. One typed outcome is
 * the honest answer because every one of them has the same remedy for the
 * adapter that sent it: fix the payload.
 */
export function parsePayload(bytes: Uint8Array): RawPayload {
  if (bytes.byteLength > MAX_PAYLOAD_BYTES) throw malformed();
  // Invalid UTF-8 decodes to U+FFFD inside strings (Go parity) and to a
  // syntax error anywhere else, because U+FFFD is not valid JSON syntax.
  const text = new TextDecoder("utf-8").decode(bytes);
  const root = parseOrderedJson(text);
  if (root === null || root.kind !== "object") throw malformed();
  const entries = root.entries;
  for (const key of REQUIRED_KEYS) {
    if (objectGet(entries, key) === undefined) throw malformed();
  }
  for (const e of entries) {
    if (!OPTIONAL_KEYS.has(e.key) && !(REQUIRED_KEYS as readonly string[]).includes(e.key)) {
      throw malformed();
    }
  }
  const schemaVersion = reqUint(entries, "schema_version", 32);
  const eventTimeMs = reqUint(entries, "event_time_ms", 64);
  return {
    schemaVersion: Number(schemaVersion),
    eventTimeMs,
    lane: reqString(entries, "lane"),
    harness: reqString(entries, "harness"),
    adapterVersion: reqString(entries, "adapter_version"),
    sessionId: reqString(entries, "session_id"),
    turnId: reqString(entries, "turn_id"),
    capturePolicy: reqString(entries, "capture_policy"),
    turnOutcome: reqString(entries, "turn_outcome"),
    userContent: reqString(entries, "user_content"),
    assistantResult: reqString(entries, "assistant_result"),
    world: optString(entries, "world"),
    scope: optString(entries, "scope"),
    workspaceRoot: optString(entries, "workspace_root"),
    branchOf: optString(entries, "branch_of"),
    host: optString(entries, "host"),
    tools: optTools(entries),
  };
}

// reqString extracts a required string field, rejecting JSON null and
// non-strings.
function reqString(entries: readonly JsonEntry[], key: string): string {
  const v = objectGet(entries, key);
  if (v === undefined || v.kind !== "string") throw malformed();
  return v.value;
}

// optString extracts an optional string field: absent or JSON null maps to
// null, anything else must be a string.
function optString(entries: readonly JsonEntry[], key: string): string | null {
  const v = objectGet(entries, key);
  if (v === undefined || v.kind === "null") return null;
  if (v.kind !== "string") throw malformed();
  return v.value;
}

// reqUint extracts a required unsigned integer field. Only bare decimal
// digits are accepted — no sign, fraction, or exponent. These fields feed
// identity and digest derivation, so each value must have exactly one
// textual form; accepting 1.0e3 for 1000 would make identity depend on
// formatting.
function reqUint(entries: readonly JsonEntry[], key: string, bitSize: 32 | 64): bigint {
  const v = objectGet(entries, key);
  if (v === undefined || v.kind !== "number") throw malformed();
  if (!/^[0-9]+$/.test(v.literal)) throw malformed();
  const n = BigInt(v.literal);
  if (n >= 1n << BigInt(bitSize)) throw malformed();
  return n;
}

// optTools extracts the optional tools array. Each element must be an
// object carrying exactly one "name" string — the closed schema does not
// admit future tool metadata without a schema version bump.
function optTools(entries: readonly JsonEntry[]): Tool[] | null {
  const v = objectGet(entries, "tools");
  if (v === undefined || v.kind === "null") return null;
  if (v.kind !== "array") throw malformed();
  const tools: Tool[] = [];
  for (const e of v.items) {
    if (e.kind !== "object" || e.entries.length !== 1) throw malformed();
    const nameV = objectGet(e.entries, "name");
    if (nameV === undefined || nameV.kind !== "string") throw malformed();
    tools.push({ name: nameV.value });
  }
  return tools;
}

/**
 * Checks a parsed payload against the closed contract and returns the
 * capture-ready Payload. The check order is fixed so that a payload with
 * several problems always reports the same first failure, which is what
 * lets an adapter test against a stable error code.
 */
export function validate(raw: RawPayload): Payload {
  const fail = (code: CaptureErrorCode): never => {
    throw new CaptureError(code);
  };
  if (raw.schemaVersion !== PAYLOAD_SCHEMA_VERSION) fail("UnsupportedSchemaVersion");
  if (raw.world === null) fail("InvalidWorld");
  if (raw.scope === null) fail("InvalidScope");
  if (!validWorld(raw.world!)) fail("InvalidWorld");
  if (!validScope(raw.scope!)) fail("InvalidScope");
  if (!validLane(raw.lane)) fail("InvalidLane");
  if (raw.eventTimeMs < MIN_EVENT_TIME_MS || raw.eventTimeMs > MAX_EVENT_TIME_MS) {
    fail("ImplausibleEventTime");
  }
  if (!validToken(raw.harness)) fail("InvalidHarness");
  if (!validToken(raw.adapterVersion)) fail("InvalidAdapterVersion");
  if (!validToken(raw.sessionId)) fail("InvalidSessionId");
  if (!validToken(raw.turnId)) fail("InvalidTurnId");
  if (!validToken(raw.capturePolicy)) fail("InvalidCapturePolicy");
  if (!validToken(raw.turnOutcome)) fail("InvalidTurnOutcome");
  if (raw.userContent.length === 0) fail("EmptyUserContent");
  if (raw.assistantResult.length === 0) fail("EmptyAssistantResult");
  if (utf8ByteLength(raw.userContent) > MAX_CONTENT_BYTES || utf8ByteLength(raw.assistantResult) > MAX_CONTENT_BYTES) {
    fail("OversizedContent");
  }
  if (!raw.userContent.isWellFormed() || !raw.assistantResult.isWellFormed()) fail("InvalidUtf8");
  const tools = raw.tools ?? [];
  if (tools.length > MAX_TOOLS) fail("TooManyTools");
  for (const t of tools) {
    if (!validToken(t.name)) fail("InvalidToolName");
  }
  if (raw.workspaceRoot !== null && !validPath(raw.workspaceRoot)) fail("InvalidWorkspaceRoot");
  if (raw.branchOf !== null && !validPath(raw.branchOf)) fail("InvalidBranchOf");
  if (raw.host !== null && !validToken(raw.host)) fail("InvalidHost");
  return {
    world: raw.world!,
    scope: raw.scope!,
    lane: raw.lane as Lane,
    harness: raw.harness,
    adapterVersion: raw.adapterVersion,
    sessionId: raw.sessionId,
    turnId: raw.turnId,
    eventTimeMs: Number(raw.eventTimeMs),
    capturePolicy: raw.capturePolicy,
    turnOutcome: raw.turnOutcome,
    userContent: raw.userContent,
    assistantResult: raw.assistantResult,
    tools,
    workspaceRoot: raw.workspaceRoot,
    branchOf: raw.branchOf,
    host: raw.host,
  };
}
